ip_proofs/src/gipa.rs

```diff
@@ -0,0 +1,556 @@
+                let ck_a_2 = &ck_a[split..];
+
+                let m_b_1 = &m_b[..split];
+                let m_b_2 = &m_b[split..];
+                let ck_b_1 = &ck_b[split..];
+                let ck_b_2 = &ck_b[..split];
+
+                let com_1 = (
+                    LMC::commit(ck_a_1, m_a_1)?,
+                    RMC::commit(ck_b_1, m_b_1)?,
+                    IPC::commit(ck_t, &vec![IP::inner_product(m_a_1, m_b_1)?])?,
+                );
+                let com_2 = (
+                    LMC::commit(ck_a_2, m_a_2)?,
+                    RMC::commit(ck_b_2, m_b_2)?,
+                    IPC::commit(ck_t, &vec![IP::inner_product(m_a_2, m_b_2)?])?,
+                );
+
+                // Fiat-Shamir challenge
+                let mut counter_nonce: usize = 0;
+                let (c, c_inv) = loop {
+                    let mut hash_input = Vec::new();
+                    hash_input.extend_from_slice(&counter_nonce.to_be_bytes()[..]);
+                    //TODO: Should use CanonicalSerialize instead of ToBytes
+                    hash_input.extend_from_slice(&to_bytes![
+                        transcript, com_1.0, com_1.1, com_1.2, com_2.0, com_2.1, com_2.2
+                    ]?);
+                    if let Some(c) = LMC::Scalar::from_random_bytes(&D::digest(&hash_input)) {
+                        if let Some(c_inv) = c.inverse() {
+                            break (c, c_inv);
+                        }
+                    };
+                    counter_nonce += 1;
+                };
+
+                // Set up values for next step of recursion
+                let m_a_recurse = m_a_1
+                    .iter()
+                    .map(|a| mul_helper(a, &c))
+                    .zip(m_a_2)
+                    .map(|(a_1, a_2)| a_1.clone() + a_2.clone())
+                    .collect::<Vec<LMC::Message>>();
+
+                let m_b_recurse = m_b_2
+                    .iter()
+                    .map(|b| mul_helper(b, &c_inv))
+                    .zip(m_b_1)
+                    .map(|(b_1, b_2)| b_1.clone() + b_2.clone())
+                    .collect::<Vec<RMC::Message>>();
+
+                let ck_a_recurse = ck_a_2
+                    .iter()
+                    .map(|a| mul_helper(a, &c_inv))
+                    .zip(ck_a_1)
+                    .map(|(a_1, a_2)| a_1.clone() + a_2.clone())
+                    .collect::<Vec<LMC::Key>>();
+
+                let ck_b_recurse = ck_b_1
+                    .iter()
+                    .map(|b| mul_helper(b, &c))
+                    .zip(ck_b_2)
+                    .map(|(b_1, b_2)| b_1.clone() + b_2.clone())
+                    .collect::<Vec<RMC::Key>>();
+
+                let (mut r_steps, r_base, mut r_transcript, ck_base) = Self::recursive_prove(
+                    (&m_a_recurse, &m_b_recurse),
+                    (&ck_a_recurse, &ck_b_recurse, ck_t),
+                    &c,
+                )?;
```

Comment:
Is there a way to do this in place/using iteration? Currently the memory required increases with the depth of recursion

Reply:
:+1: 

Yes, I've made this change now for all of the recursive implementations in `gipa`.